File: automation-js/src/session.ts
/**
 * Console/network/error capture + markdown extraction for a Playwright
 * (Node) page — the JS/TS twin of the Python `automation/` package's
 * `ExtractorSession` / `AsyncExtractorSession`. One class here covers both,
 * since Node's Playwright API is async-only to begin with.
 *
 * Works with anything that hands you a real Playwright `Page`: `playwright`,
 * `playwright-core`, `@playwright/test`, and `camoufox-js` (which wraps
 * `playwright-core` directly).
 *
 * Design notes (see the repo README for the full rationale — same as the
 * Python package):
 *   - Console/network/page-error capture uses Playwright's *native* event
 *     hooks (`page.on('console'/'request'/'requestfinished'/'requestfailed'/'pageerror')`)
 *     rather than an injected JS patcher, because:
 *       1. It's more robust — no reliance on `page.addInitScript()`, which is
 *          known to be unreliable under Camoufox's isolated-world execution
 *          model (https://github.com/daijro/camoufox/issues/48).
 *       2. It captures more than the extension does (any resource type, not
 *          just fetch/XHR), and can't be blocked by a page's CSP.
 *   - Markdown extraction and CSS-selector helpers *do* need to run inside
 *     the page (DOM traversal). Those live in `extension/core/*.js` and are
 *     read from disk here, then run via `page.evaluate()`. That's safe under
 *     Camoufox's default isolated world because that code only reads the DOM
 *     — it never writes to the live page — so no `mainWorld`/init-script
 *     workaround is required.
 */

import type { ConsoleMessage, Page, Request } from 'playwright-core';
import { readSharedJs } from './loadJs.js';

const MAX_ENTRIES = 200;

export interface ConsoleEntry {
  level: string;
  ts: number;
  msg: string;
}

export interface ErrorEntry {
  type: string;
  ts: number;
  msg: string;
  source: string;
  line: number;
  col: number;
  stack: string;
}

export interface NetworkEntry {
  type: string;
  method: string;
  url: string;
  status: number;
  ts: number;
  duration: number;
  error: string | null;
}

export interface CaptureStore {
  console: ConsoleEntry[];
  errors: ErrorEntry[];
  network: NetworkEntry[];
}

export interface ExtractedMarkdown {
  url: string;
  title: string;
  ts: number;
  selector: string;
  markdown: string;
}

const CONSOLE_LEVEL_MAP: Record<string, string> = { warning: 'warn' };

function now(): number {
  return Date.now();
}

class RingStore implements CaptureStore {
  console: ConsoleEntry[] = [];
  errors: ErrorEntry[] = [];
  network: NetworkEntry[] = [];

  push<K extends keyof CaptureStore>(which: K, entry: CaptureStore[K][number]): void {
    const arr = this[which];
    (arr as Array<CaptureStore[K][number]>).push(entry);
    while (arr.length > MAX_ENTRIES) arr.shift();
  }

  clear(which?: keyof CaptureStore): void {
    if (which) {
      this[which].length = 0;
    } else {
      this.console.length = 0;
      this.errors.length = 0;
      this.network.length = 0;
    }
  }

  asDict(): CaptureStore {
    return { console: this.console, errors: this.errors, network: this.network };
  }
}

function buildExtractScript(selector: string | null | undefined): string {
  const domJs = readSharedJs('dom.js');
  const selJson = JSON.stringify(selector || '');
  return `
(() => {
${domJs}
let el = null;
const sel = ${selJson};
if (sel) { try { el = document.querySelector(sel); } catch (_) { el = null; } }
if (!el) el = document.body;
return extractMarkdown(el);
})()
`;
}

function buildPromptScript(
  meta: Pick<ExtractedMarkdown, 'url' | 'title' | 'ts' | 'selector'>,
  markdown: string,
  store: CaptureStore,
  maxChars: number | undefined,
): string {
  const promptJs = readSharedJs('prompt.js');
  const maxCharsJs = maxChars === undefined ? 'undefined' : JSON.stringify(maxChars);
  return `
(() => {
${promptJs}
return buildAIPrompt(${JSON.stringify(meta)}, ${JSON.stringify(markdown)}, ${JSON.stringify(store)}, ${maxCharsJs});
})()
`;
}

/**
 * Attach to a page BEFORE navigating, so no console/network/error activity
 * from the initial load is missed.
 *
 * @example
 * ```ts
 * import { chromium } from 'playwright-core';
 * import { ExtractorSession } from 'context-extractor';
 *
 * const page = await (await chromium.launch()).newPage();
 * const session = new ExtractorSession(page);
 * await page.goto('https://example.com', { waitUntil: 'networkidle' });
 * console.log(await session.buildAiPrompt());       // or session.extractMarkdown('#main')
 * ```
 */
export class ExtractorSession {
  readonly page: Page;
  private readonly store = new RingStore();
  private readonly requestStarts = new WeakMap<Request, number>();

  constructor(page: Page) {
    this.page = page;
    page.on('console', this.onConsole);
    page.on('pageerror', this.onPageError);
    page.on('request', this.onRequest);
    page.on('requestfinished', this.onRequestFinished);
    page.on('requestfailed', this.onRequestFailed);
  }

  /** Stop listening. Call this if you keep a page open long after you're done capturing. */
  detach(): void {
    this.page.off('console', this.onConsole);
    this.page.off('pageerror', this.onPageError);
    this.page.off('request', this.onRequest);
    this.page.off('requestfinished', this.onRequestFinished);
    this.page.off('requestfailed', this.onRequestFailed);
  }

  getStore(): CaptureStore {
    return this.store.asDict();
  }

  clear(which?: keyof CaptureStore): void {
    this.store.clear(which);
  }

  async extractMarkdown(selector?: string | null): Promise<ExtractedMarkdown> {
    const markdown = (await this.page.evaluate(buildExtractScript(selector))) as string;
    return {
      url: this.page.url(),
      title: await this.page.title(),
      ts: now(),
      selector: selector || 'body',
      markdown: markdown || '',
    };
  }

  async buildAiPrompt(selector?: string | null, maxChars?: number): Promise<string> {
    const extracted = await this.extractMarkdown(selector);
    const meta = {
      url: extracted.url,
      title: extracted.title,
      ts: extracted.ts,
      selector: extracted.selector,
    };
    const script = buildPromptScript(meta, extracted.markdown, this.getStore(), maxChars);
    return (await this.page.evaluate(script)) as string;
  }

  private onConsole = (msg: ConsoleMessage): void => {
    const type = msg.type();
    const level = CONSOLE_LEVEL_MAP[type] ?? type;
    this.store.push('console', { level, ts: now(), msg: msg.text() });
  };

  private onPageError = (error: Error): void => {
    this.store.push('errors', {
      type: 'error',
      ts: now(),
      msg: error?.message ?? String(error),
      source: '',
      line: 0,
      col: 0,
      stack: error?.stack ?? '',
    });
  };

  private onRequest = (request: Request): void => {
    this.requestStarts.set(request, Date.now());
  };

  private onRequestFinished = async (request: Request): Promise<void> => {
    const start = this.requestStarts.get(request);
    this.requestStarts.delete(request);
    const duration = start ? Date.now() - start : 0;
    let status = 0;
    try {
      const response = await request.response();
      if (response) status = response.status();
    } catch {
      /* response unavailable — leave status 0 */
    }
    this.store.push('network', {
      type: request.resourceType(),
      method: request.method(),
      url: request.url(),
      status,
      ts: now(),
      duration,
      error: null,
    });
  };

  private onRequestFailed = (request: Request): void => {
    const start = this.requestStarts.get(request);
    this.requestStarts.delete(request);
    const duration = start ? Date.now() - start : 0;
    const failure = request.failure();
    this.store.push('network', {
      type: request.resourceType(),
      method: request.method(),
      url: request.url(),
      status: 0,
      ts: now(),
      duration,
      error: failure?.errorText || 'Network error',
    });
  };
}

File: automation/context_extractor/session.py
"""Console/network/error capture + markdown extraction for a Playwright page.

Works with:
  - Plain Playwright (any browser: chromium, firefox, webkit)
  - Camoufox (daijro/camoufox), sync or async — Camoufox hands you a normal
    Playwright Page/BrowserContext, so everything here works unmodified.

Design notes (see project README for the full rationale):
  - Console/network/page-error capture uses Playwright's *native* event hooks
    (page.on("console"/"request"/"requestfinished"/"requestfailed"/"pageerror")).
    This deliberately avoids injecting a JS patcher into the page (the trick
    the browser extension has to use), because:
      1. It's more robust — no reliance on page.add_init_script(), which is
         known to be unreliable under Camoufox's isolated-world execution
         model (see https://github.com/daijro/camoufox/issues/48).
      2. It captures more than the extension does (any resource type, not
         just fetch/XHR), and can't be blocked by a page's CSP.
  - Markdown extraction and CSS-selector helpers *do* need to run inside the
    page (DOM traversal). Those live in extension/core/*.js and are read from
    disk here, then run via page.evaluate(). This is safe under Camoufox's
    default *isolated* world because that code only reads the DOM / mutates a
    detached clone — it never writes to the live page — so no
    `main_world_eval` workaround is required.
"""

from __future__ import annotations

import json
import time
from pathlib import Path
from typing import Any, Optional

_JS_DIR = Path(__file__).parent / "js"
MAX_ENTRIES = 200


def _read_js(name: str) -> str:
    path = _JS_DIR / name
    if not path.exists():
        raise FileNotFoundError(
            f"Missing shared JS file: {path}. This package expects to run from "
            "a checkout of the context-extractor repo where automation/context_extractor/js/*.js "
            "are symlinks into extension/core/."
        )
    return path.read_text(encoding="utf-8")


def _now_ms() -> int:
    return int(time.time() * 1000)


class _Store:
    """Bounded ring-buffer store matching the extension's shape exactly, so
    build_ai_prompt() output is identical between the extension and here."""

    def __init__(self) -> None:
        self.console: list[dict[str, Any]] = []
        self.errors: list[dict[str, Any]] = []
        self.network: list[dict[str, Any]] = []

    def push(self, which: str, entry: dict[str, Any]) -> None:
        arr = getattr(self, which)
        arr.append(entry)
        while len(arr) > MAX_ENTRIES:
            arr.pop(0)

    def clear(self, which: Optional[str] = None) -> None:
        if which:
            getattr(self, which).clear()
        else:
            self.console.clear()
            self.errors.clear()
            self.network.clear()

    def as_dict(self) -> dict[str, Any]:
        return {"console": self.console, "errors": self.errors, "network": self.network}


_CONSOLE_LEVEL_MAP = {"warning": "warn"}


def _build_extract_script(selector: Optional[str]) -> str:
    dom_js = _read_js("dom.js")
    sel_json = json.dumps(selector or "")
    return f"""
(() => {{
{dom_js}
let el = null;
const sel = {sel_json};
if (sel) {{ try {{ el = document.querySelector(sel); }} catch (_) {{ el = null; }} }}
if (!el) el = document.body;
return extractMarkdown(el);
}})()
"""


def _build_prompt_script(
    meta: dict[str, Any], markdown: str, store: dict[str, Any], max_chars: Optional[int] = None
) -> str:
    prompt_js = _read_js("prompt.js")
    max_chars_js = json.dumps(max_chars) if max_chars is not None else "undefined"
    return f"""
(() => {{
{prompt_js}
return buildAIPrompt({json.dumps(meta)}, {json.dumps(markdown)}, {json.dumps(store)}, {max_chars_js});
}})()
"""


class _CaptureMixin:
    """Shared event-handling logic. Playwright event callbacks are plain
    synchronous functions in both the sync and async APIs (Playwright invokes
    them itself; you never await them), so this is safe to share as-is."""

    def _init_capture_state(self) -> None:
        self.store = _Store()
        self._request_starts: dict[int, float] = {}

    def _on_console(self, msg) -> None:
        level = _CONSOLE_LEVEL_MAP.get(msg.type, msg.type)
        self.store.push("console", {"level": level, "ts": _now_ms(), "msg": msg.text})

    def _on_pageerror(self, error) -> None:
        message = getattr(error, "message", None) or str(error)
        stack = getattr(error, "stack", "") or ""
        self.store.push("errors", {
            "type": "error", "ts": _now_ms(), "msg": message,
            "source": "", "line": 0, "col": 0, "stack": stack,
        })

    def _on_request(self, request) -> None:
        self._request_starts[id(request)] = time.monotonic()

    def _on_requestfinished(self, request) -> None:
        start = self._request_starts.pop(id(request), None)
        duration = round((time.monotonic() - start) * 1000) if start else 0
        status = 0
        try:
            response = request.response()
            if response:
                status = response.status
        except Exception:
            pass
        self.store.push("network", {
            "type": request.resource_type, "method": request.method, "url": request.url,
            "status": status, "ts": _now_ms(), "duration": duration, "error": None,
        })

    def _on_requestfailed(self, request) -> None:
        start = self._request_starts.pop(id(request), None)
        duration = round((time.monotonic() - start) * 1000) if start else 0
        failure = getattr(request, "failure", None)
        error_text = failure.get("errorText") if isinstance(failure, dict) else str(failure or "Network error")
        self.store.push("network", {
            "type": request.resource_type, "method": request.method, "url": request.url,
            "status": 0, "ts": _now_ms(), "duration": duration, "error": error_text or "Network error",
        })

    def get_store(self) -> dict[str, Any]:
        return self.store.as_dict()

    def clear(self, which: Optional[str] = None) -> None:
        self.store.clear(which)


class ExtractorSession(_CaptureMixin):
    """Sync API. Use with `playwright.sync_api` or sync `camoufox.Camoufox`.

    Example:
        from playwright.sync_api import sync_playwright
        from context_extractor import ExtractorSession

        with sync_playwright() as p:
            page = p.chromium.launch().new_page()
            session = ExtractorSession(page)
            page.goto("https://example.com")
            print(session.build_ai_prompt())
    """

    def __init__(self, page) -> None:
        self.page = page
        self._init_capture_state()
        page.on("console", self._on_console)
        page.on("pageerror", self._on_pageerror)
        page.on("request", self._on_request)
        page.on("requestfinished", self._on_requestfinished)
        page.on("requestfailed", self._on_requestfailed)

    def extract_markdown(self, selector: Optional[str] = None) -> dict[str, Any]:
        markdown = self.page.evaluate(_build_extract_script(selector))
        return {
            "url": self.page.url,
            "title": self.page.title(),
            "ts": _now_ms(),
            "selector": selector or "body",
            "markdown": markdown or "",
        }

    def build_ai_prompt(self, selector: Optional[str] = None, max_chars: Optional[int] = None) -> str:
        extracted = self.extract_markdown(selector)
        meta = {k: extracted[k] for k in ("url", "title", "ts", "selector")}
        script = _build_prompt_script(meta, extracted["markdown"], self.get_store(), max_chars)
        return self.page.evaluate(script)


class AsyncExtractorSession(_CaptureMixin):
    """Async API. Use with `playwright.async_api` or `camoufox.AsyncCamoufox`.

    Example:
        from camoufox.async_api import AsyncCamoufox
        from context_extractor import AsyncExtractorSession

        async with AsyncCamoufox(headless=True) as browser:
            page = await browser.new_page()
            session = AsyncExtractorSession(page)
            await page.goto("https://example.com")
            print(await session.build_ai_prompt())
    """

    def __init__(self, page) -> None:
        self.page = page
        self._init_capture_state()
        page.on("console", self._on_console)
        page.on("pageerror", self._on_pageerror)
        page.on("request", self._on_request)
        page.on("requestfinished", self._on_requestfinished)
        page.on("requestfailed", self._on_requestfailed)

    async def extract_markdown(self, selector: Optional[str] = None) -> dict[str, Any]:
        markdown = await self.page.evaluate(_build_extract_script(selector))
        return {
            "url": self.page.url,
            "title": await self.page.title(),
            "ts": _now_ms(),
            "selector": selector or "body",
            "markdown": markdown or "",
        }

    async def build_ai_prompt(self, selector: Optional[str] = None, max_chars: Optional[int] = None) -> str:
        extracted = await self.extract_markdown(selector)
        meta = {k: extracted[k] for k in ("url", "title", "ts", "selector")}
        script = _build_prompt_script(meta, extracted["markdown"], self.get_store(), max_chars)
        return await self.page.evaluate(script)
